### PR TITLE
Whisper streaming + VAD configuration (auditok as default instead of silero)

### PR DESCRIPTION
**Pull request by Jeronymous**

### CODE REVIEW

**Comment by Jeronymous** [comment]
@AudranBert Don't forget to update the README about streaming, comparing what's in kaldi/README.md

For instance websocket part and http with ENABLE_STREAMING needs to be documented in whipser/README.md

**Comment by AudranBert** [comment]
> @AudranBert Don't forget to update the README about streaming, comparing what's in kaldi/README.md
> 
> For instance websocket part and http with ENABLE_STREAMING needs to be documented in whipser/README.md

I updated the REAME [in this commit](https://github.com/linto-ai/linto-stt/pull/41/commits/e6e993489e519c0b6182c58847593613eae5c9f3)

File: whisper/.envdefault [hunk 1]
```diff
@@ -40,7 +40,15 @@ PROMPT=
...
 # Number of workers
 CONCURRENCY=2
+
+# WEBSOCKET PARAMETERS
+STREAMING_PORT=80
+
+# HTTP PARAMETERS
+ENABLE_STREAMING=true
+
+USE_VAD=auditok
```
Comment (Jeronymous) [inline whisper/.envdefault:54]: This should go in the section "STT MODELING PARAMETERS", with a comment explaining what it is about, and what are possible values.

**Review by Jeronymous — COMMENTED** [review]
Concerning USE_VAD, shouldn't we use just "VAD" given that it's not just a boolean variable (it can take the name of an explicit method).

Also, this seems to be implemented for streaming only for now.
Do you plan to integrate it properly with offline decoding (namely faster_whisper with auditok)?

**Comment by AudranBert** [comment]
- Streaming is working in http and websocket mode
- The VAD has been refactored for decoding and streaming
- Readme updated
- Fixed workers on CPU using a gunicorn hook
Anything left?

File: whisper/stt/processing/vad.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+        )
+    segments = [(seg["start"], seg["end"]) for seg in segments]
+    if len(segments) == 0:
+        if avoid_empty_speech:
+            segments = [(0, audio.shape[-1])]
+        else:
+            np.array([]), [], lambda t, t2=None: t if t2 is None else [t, t2]
+
+    audio_speech = np.concatenate([audio[..., s:e] for s, e in segments], axis=-1)
+    # audio_speech = torch.cat([audio[..., s:e] for s,e in segments], dim=-1)
```
Comment (Jeronymous) [inline whisper/stt/processing/vad.py:73]: remove the commented out line.

And move this below the "`if return_format == "dict"`".
It's better to avoid computing this when possible (to save RAM and time)

File: whisper/.envdefault [hunk 3]
```diff
@@ -40,7 +40,15 @@ PROMPT=
...
 # Number of workers
 CONCURRENCY=2
+
+# WEBSOCKET PARAMETERS
+STREAMING_PORT=80
+
+# HTTP PARAMETERS
+ENABLE_STREAMING=true
+
+VAD=auditok
```
Comment (Jeronymous) [inline whisper/.envdefault:54]: Please move this up, in the relevant section (it's not about serving parameters, it's more related to the details of the STT)

File: whisper/README.md [hunk 4]
```diff
@@ -114,17 +114,22 @@ cp whisper/.envdefault whisper/.env
...
+| LANGUAGE | Language to recognize | `*` \| `fr` \| `fr-FR` \| `French` \| `en` \| `en-US` \| `English` \| ... |
+| PROMPT | Prompt to use for the Whisper model | `some free text to encourage a certain transcription style (disfluencies, no punctuation, ...)` |
+| DEVICE | Device to use for the model (by default, GPU/CUDA is used if it is available, CPU otherwise) | `cpu` \| `cuda` |
+| NUM_THREADS | Number of threads (maximum) to use for things running on CPU | `1` \| `4` \| ... |
+| CUDA_VISIBLE_DEVICES | GPU device index to use, when running on GPU/CUDA. We also recommend to set `CUDA_DEVICE_ORDER=PCI_BUS_ID` on multi-GPU machines | `0` \| `1` \| `2` \| ... |
 | CONCURRENCY | Maximum number of parallel requests | `2` |
-| SERVICE_NAME | (For the task mode) queue's name for task processing | `my-stt` |
-| SERVICE_BROKER | (For the task mode) URL of the message broker | `redis://my-broker:6379` |
+| VAD | Activate (and specify which method) Voice Activity Detection for removing non speech segments | `if the argument is not specidifed, it will use auditok` \| `true (will use auditok)` \| `false` \| `auditok` \| `silero` \|
```
Comment (Jeronymous) [inline whisper/README.md:125]: > if the argument is not specidifed, it will use auditok

move this in the second column (in a parenthesis).
Also you can simplify and fix the typo.

I would just put in the second column:
"Voice Activity Detection method. Use "false" to disable. If not specified, the default is auditok VAD."

and for the last column:
```
`true` \| `false` \| `auditok` \| `silero`
```

BTW, do 0/1 work also (as true/false)?



The last column is only for possible values, when specified

File: whisper/stt/processing/vad.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    return result
+
+
+def get_vad_segments(
+    audio,
+    sample_rate=16000,
+    output_sample=False,
+    min_speech_duration=0.1,
+    min_silence_duration=0.1,
+    dilatation=0.5,
```
Comment (Jeronymous) [inline whisper/stt/processing/vad.py:139]: Those could be parameterizable as environment variables as well
(VAD_DILATATION, ...)

File: whisper/stt/__init__.py [hunk 6]
```diff
@@ -12,6 +12,16 @@
 # see https://github.com/guillaumekln/faster-whisper/issues/150
 os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"  # GPU in the right order
 
+if os.environ.get("VAD","auditok") in [True, "true", 1]:
```
Comment (Jeronymous) [inline whisper/stt/__init__.py:15]: You will never goot a boolean or an integer with os.environ.
It should be `os.environ.get("VAD","auditok").lower() in ["true", "1"]`
(same for False).

Also you should check here that the specified VAD method is correct.
To avoid only throwing at the first call.
(if the configuration is not right, we have to fail early)

File: whisper/stt/processing/__init__.py [hunk 7]
```diff
@@ -91,7 +84,18 @@ def __call__(self, *args, **kwargs):
...
-MODEL = (model, alignment_model)
+# Load ASR model
+model_type = os.environ.get("MODEL", "medium")
+logger.info(
+    f"Loading Whisper model {model_type} ({'local' if os.path.exists(model_type) else 'remote'})..."
+)
+try:
+    model = LazyLoadedModel(model_type, device=device, num_threads=NUM_THREADS)
+    MODEL = (model, alignment_model)
+    if str(device).lower() != "cpu":
```
Comment (Jeronymous) [inline whisper/stt/processing/__init__.py:95]: This test is a bit shaky.
Please use USE_GPU.
Also add a comment to explain why the warmup is done at the start with GPU.

File: whisper/stt/processing/vad.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+                    else "snakers4/silero-vad"
+                )
+                source = "github"
+            if need_folder_hack:
+                apply_folder_hack()
+            try:
+                if _vad_import is None:
+                    from torch.hub import load as torch_load
+
+                    _vad_import = torch_load
```
Comment (Jeronymous) [inline whisper/stt/processing/vad.py:232]: No need for that

**Comment by AudranBert** [comment]
Kaldi offline and streaming transcription worked

**Comment by AudranBert** [comment]
👍
